server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/mail/EmailBodyPart.scala

```diff
@@ -19,11 +19,102 @@
 
 package org.apache.james.jmap.mail
 
+import java.io.OutputStream
+
+import cats.implicits._
+import com.google.common.io.CountingOutputStream
+import eu.timepit.refined.api.Refined
+import eu.timepit.refined.auto._
 import eu.timepit.refined.numeric.NonNegative
+import eu.timepit.refined.refineV
 import org.apache.james.jmap.mail.Email.Size
+import org.apache.james.jmap.mail.PartId.PartIdValue
+import org.apache.james.mailbox.model.MessageId
+import org.apache.james.mime4j.dom.{Entity, Message, Multipart}
+import org.apache.james.mime4j.message.DefaultMessageWriter
+
+import scala.jdk.CollectionConverters._
+import scala.util.{Failure, Success, Try}
+
+object PartId {
+  type PartIdValue = Int Refined NonNegative
+}
+
+case class PartId(value: PartIdValue) {
+  def serialize: String = value.toString
+
+  def next: PartId = refineV[NonNegative](value + 1) match {
+    case scala.Right(incrementedValue) => PartId(incrementedValue)
+    case Left(e) => throw new IllegalArgumentException(e)
+  }
+}
+
+object EmailBodyPart {
+  def of(messageId: MessageId, message: Message): Try[EmailBodyPart] =
+    of(messageId, PartId(1), message).map(_._1)
 
-case class PartId(value: NonNegative) {
-  def serialize(): String = value.toString
+  def of(messageId: MessageId, partId: PartId, entity: Entity): Try[(EmailBodyPart, PartId)] =
+    entity.getBody match {
+      case multipart: Multipart =>
+        val scanResults: Try[List[(Option[EmailBodyPart], PartId)]] = multipart.getBodyParts
+          .asScala.toList
+          .scanLeft[Try[(Option[EmailBodyPart], PartId)]](Success((None, partId)))(traverse(messageId))
+          .sequence
+        val highestPartIdValidation: Try[PartId] = scanResults.map(list => list.map(_._2).reverse.headOption.getOrElse(partId))
+        val childrenValidation: Try[List[EmailBodyPart]] = scanResults.map(list => list.flatMap(_._1))
+
+        zip(childrenValidation, highestPartIdValidation)
+            .flatMap({
```

Comment:
when calling function with a lambda block parameter you don't need '({' only '{'